contracts/AccessLogs.sol

```diff
@@ -0,0 +1,165 @@
+// Copyright 2023 Cartesi Pte. Ltd.
+
+// SPDX-License-Identifier: Apache-2.0
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+pragma solidity ^0.8.0;
+
+import "./interfaces/IAccessLogs.sol";
+import "./Memory.sol";
+import "./UArchCompat.sol";
+
+library AccessLogs {
+    using Memory for Memory.AlignedSize;
+
+    //
+    // Read methods
+    //
+
+    function readRegion(
+        IAccessLogs.Context memory a,
+        Memory.Region memory region
+    ) internal pure returns (bytes32) {
+        bytes32 drive = a.hashes[a.currentHashIndex++];
+        (bytes32 rootHash, uint8 siblingCount) = getRoot(
+            region,
+            drive,
+            a.hashes,
+            a.currentHashIndex
+        );
+        a.currentHashIndex += siblingCount;
+
+        require(
+            a.currentRootHash == rootHash,
+            "Read region root doesn't match"
+        );
+
+        return drive;
+    }
+
+    function readLeaf(
+        IAccessLogs.Context memory a,
+        uint64 readAddress
```

Comment:
I believe this should be the index of the leaf, rather than the memory address. A leaf (and its index) represents a word; an address represents a byte.

I also think the type here could be `Memory.Stride`.